src/label_producer.py

```diff
@@ -0,0 +1,205 @@
+        
+        Returns:
+            The frequency of the term from the frequencies dict. 0 if it doesn't
+            exist in th the dict
+        """
+        return self.__TERM_FREQUENCIES_DICT.get(term, 0)
+    
+    def __get_tf_idf(self, term, count):
+        """
+        Calculate the tf/idf of the given term
+        
+        Args:
+            term: the term to calculate tf/idf of
+            count: the term frequency
+            
+        Returns:
+            The tf/idf value for the given term
+        """
+        hit_docs = self.__get_term_frequency(term)
+        if hit_docs == 0:
+          return 0
+        idf = math.log(self.__TOTAL_ARTICLES / hit_docs)
+        tf_idf = count * idf
+        return tf_idf
+
+    def __filter_words(self, text, stop_words):
+        """
+        Filter out words from the given text
+        
+        Args:
+            text: the text to filter
+            stop_words: a list of words to filter out
+        
+        Returns:
+            All the words in the text that are not in the stop_words list, have
+            length greater than 1 and are alphabetical characters
+        """
+        return [word for word in text if word not in stop_words and len(word) > 1 and word.isalpha()]
+    
+    def __tokenize(self, text, stop_words):
+        """
+        Tokenize the fiven text
+        
+        Args:
+            text: the text to tokenize
+            stop_words: a list of words to filter out
+            
+        Returns:
+            text after tokenization, conversion to lowercase and filtering of words
+            using the filterWords function
+        """
+        return self.__filter_words(nltk.word_tokenize(text), stop_words)
+    
+    def __term_counter(self, term, stop_words):
+        """
+        Count the occurrences of different words in the sentences that contain
+        all the words of the given term
+        
+        Args:
+            term: the term to query for
+            stop_words: a list of words to filter out
+            
+        Returns:
+            A counter of occurrences of all the unfiltered words that returned from
+            the given query
+        """
+        res = self.__get_pages(term)
+        result_words = []
+        for doc in res:
+            result_words += self.__tokenize(doc.lower(), stop_words)
+        counter = Counter(result_words)
+        return counter
+    
+    def calculate_most_probable_relations(self, terms, topn=20):
+        """
+        Get the most probable labels for the given list of terms
+        
+        Args:
+            terms: a list of 2-tuples which are terms to search labels for
+            topn: number of labels to return
+            
+        Returns:
+            A list of 'topn' labels which represent the relations between the words
+            in the terms sorted in descending order of probablity
+        """
+    
+        nltk.download('stopwords') 
```

Comment:
Extracting the terms frequency loading to a method may simplify the function

Reply:
Possibly, we want to move the frequencies and stop-words into the constructor. This way they can be loaded just once (in case of multiple calls to ```calculate_most_probable_relations```) and it'll help us avoid passing _stop_words_ to internal methods.